fix: add prefill metrics support for TensorRT-LLM disaggregated mode

keivenchang opened the pull request:
#### Overview:

Adds prefill metrics support for TensorRT-LLM disaggregated mode. Previously, only agg metrics were validated; now both prefill and decode workers expose metrics on separate ports (8081 and 8082).

#### Details:

- **KV Cache Config Fix**: Modified `main.py` to preserve `cache_transceiver_config` from YAML when adding `event_buffer_max_size`, ensuring disaggregated mode coordination works correctly
- **New Launch Script**: Added `disagg_same_gpu.sh` for single-GPU disaggregated testing
- **Dual Metrics Support**: Both prefill and decode workers now publish metrics with `--publish-events-and-metrics` flag
- **Test Enhancement**: Updated `test_trtllm.py` to validate metrics from both ports (8081 for prefill, 8082 for decode)
- **Test Infrastructure**: Extended `metric_payload_default()` to accept port parameter for multi-port metrics testing

#### Where should the reviewer start?

- `components/src/dynamo/trtllm/main.py` - KV cache config preservation logic
- `components/backends/trtllm/launch/disagg_same_gpu.sh` - new single-GPU disaggregated launch script with metrics
- `tests/serve/test_trtllm.py` - disaggregated_same_gpu test configuration with dual metrics validation

#### Related Issues:

DIS-911

/coderabbit profile chill

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Disaggregated inference now supports single-GPU execution with integrated GPU memory validation to ensure sufficient resources are available before inference begins.

* **Infrastructure**
  * Enhanced the metrics collection system to support multiple backend metric endpoints for improved visibility into system performance.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
fix: add prefill metrics support for TensorRT-LLM disaggregated mode

- Fix KV cache config to preserve cache_transceiver_config from YAML
- Add disagg_same_gpu.sh launch script with metrics on ports 8081/8082
- Add trtllm-small recipe configs for memory-optimized testing (0.24 GPU fraction)
- Add disaggregated_same_gpu test config with metrics validation
- Add port parameter to metric_payload_default for multi-port testing
- Replace bc with bash arithmetic for better portability

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

tedzhouhk commented on recipes/qwen3/trtllm-small/agg.yaml:
```diff
@@ -0,0 +1,34 @@
+moe_expert_parallel_size: 1
+enable_attention_dp: false
+max_num_tokens: 1024
+max_batch_size: 4
```
seems too low for general use cases? but please ignore me if recipe folder doesn't need to be high-performance in general. same for bs and mnt in disagg

tedzhouhk commented on recipes/qwen3/trtllm-small/agg.yaml:
```diff
@@ -0,0 +1,34 @@
+enable_chunked_prefill: true
+
+kv_cache_config:
+  free_gpu_memory_fraction: 0.24
```
seems too small, but also please ignore me if recipe folder doesn't need to be high-performance in general. same fore below mem frac in disagg

keivenchang commented on components/backends/trtllm/launch/disagg_same_gpu.sh:
```diff
@@ -0,0 +1,74 @@
+    echo "Error: Insufficient GPU memory. Required: ${REQUIRED_GB}GB, Available: ${FREE_GPU_GB}GB"
+    echo "Please free up GPU memory before running disaggregated mode on single GPU."
+    exit 1
+fi
```
Addressed.

keivenchang committed:
Merge branch 'main' into keivenchang/DIS-911__add-prefill-metrics-TRTLLM-part-2

keivenchang committed:
move trtllm test configs from recipes to tests/serve/configs

Relocate memory-optimized test configs to tests/serve/configs/trtllm/ for better organization alongside other test configurations

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
update disagg_same_gpu.sh to use new config paths and lower GPU requirement

- Update PREFILL_ENGINE_ARGS and DECODE_ENGINE_ARGS to point to tests/serve/configs/trtllm/
- Lower REQUIRED_GB from 20 to 16 for more flexible testing

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

nv-kmcgill53 reviewed (commented):
I have the same concerns as other's regarding the test yaml files in the recipes folder. I'll wait for that to be resolved before I approve

keivenchang commented on tests/serve/test_trtllm.py:
```diff
@@ -60,6 +60,20 @@ class TRTLLMConfig(EngineConfig):
             completion_payload_default(),
         ],
     ),
+    "disaggregated_same_gpu": TRTLLMConfig(
```
Search for "e2e and trtllm_marker and gpu_1 and not slow" in container-validation-backends.yml. 

Example: https://github.com/ai-dynamo/dynamo/actions/runs/19050933502/job/54423846998?pr=3983
```
tests/serve/test_trtllm.py::test_deployment[disaggregated_same_gpu] PASSED [ 77%]
```

nv-kmcgill53 replied:
Thanks!
